### PR TITLE
feat: add generic project links

### LINKED ISSUE #4320: Generic dashboard links
# Feature request

From Ant:
```
When writing documentation or providing customer support it would be useful to give someone a link directly to a certain page on their dashboard, without knowing their project reference,

it would be useful to set up some endpoints e.g. https://app.supabase.io/project/[ref]/database/backups that forwards to that page for their default (or most recent?) project
```

From Thor:
```
This is what firebase does and I quite like it:

Link without the project ID e.g. https://app.supabase.io/project/auth/templates shows a list of your projects
When you click the project it then goes to the above link with your project ID filled in https://app.supabase.io/project/zdtdtxajzydjqzuktnqx/auth/templates
I think that should we quite doable, might look into this myself if that's okay?
```

### PR DESCRIPTION
**Pull request by thorwebdev**
Closes #4320 

## What kind of change does this PR introduce?

Introduces a new generic `/settings` route that shows a list of your active projects and redirects to the relevant setting for the project that was selected.

[Update (Joshen)]
- Generic links will use `/project/_/{{slug}}` as the route structure instead of `/settings/{{slug}}`

## What is the current behavior?

Currently we don't have generic links

## What is the new behavior?

Generic settings links we can use throughout the docs and support \o/

## Additional context

image

### CODE REVIEW

**Review by DukeManh — COMMENTED** [review]
Kinda sad that I just figured out a working solution 🥲 😋.

Can I make one suggestion? This could be through [Next redirects config](https://nextjs.org/docs/api-reference/next.config.js/redirects#header-cookie-and-query-matching) and we can save the destination in the URL query and append it to the link of the project card.

```js
  async redirects() {
    return [
      {
        source: '/project/ref/:path(.{1,})',
        destination: '/?redirect=:path*',
        permanent: true,
      },
      {
        source: '/project/ref/',
        destination: '/',
        permanent: true,
      },
    ];
  }
```

Wdyt? Let me know.

**Comment by thorwebdev** [comment]
@DukeManh how do you present the UI with the users projects though?

File: studio/components/interfaces/Home/ProjectList/ProjectCard.tsx [hunk 1]
```diff
@@ -5,14 +5,15 @@ import { Project } from 'types'
 
 interface Props {
   project: Project
+  rewriteHref?: string
```
Comment (joshenlim) [inline studio/components/interfaces/Home/ProjectList/ProjectCard.tsx:8]: maybe just `href?: string`

**Comment by DukeManh** [comment]
@thorwebdev, The config above would direct a URL like '/project/ref/auth/templates' to '/?redirect=auth/templates' which shows a list of projects because it's the main page with query. Then we could add extract the `redirect` query and add it to the link of the project card, which leads users to the wanted URL.

**Comment by joshenlim** [comment]
i'll push some fixes but quick note - in this case we can't use the project layout then as the links in the left most side nav are project specific. Maybe hold this PR first, will think about the layout

**Comment by joshenlim** [comment]
cc @MildTomato as well if you have any input on how this page should look

**Comment by joshenlim** [comment]
btw i can't seem to checkout this branch properly, wondering if you guys might know why

image

**Comment by joshenlim** [comment]
this is how it'll look like now - not the best tbh, will iterate over it again

image

qq - the generic link should technically work for routes outside of auth too right? e.g /database/{{page}} or /editor

**Comment by MildTomato** [comment]
https://github.com/supabase/supabase/issues/4320#issuecomment-1046984298

If this was the chosen solution for this, we should isolate the project list, and not have the nav shown.
Does this work for all our routes btw?

**Comment by joshenlim** [comment]
okay managed to dig through firebase to get some prior art

this is how firebase does their generic links
https://console.firebase.google.com/u/1/project/_/firestore

it'll bring you to a page that lists all projects and prompt the user to choose which project to continue

clicking on the project card leads to the project with that url
e.g https://console.firebase.google.com/u/1/project/{projectRef}/firestore

i'm thinking should do something similar - the keyword "settings" before the slug is a little misleading i feel?
we could probably do the same strategy: `/project/_/auth/policies`

**Comment by joshenlim** [comment]
Okay i think this should be good to go now, main changes include

- Updating the layout to focus on the sole CTA of "choosing your project to proceed"
- Changed the route for generic project pages to `.../project/_/{{routeSlug}}` instead of `.../settings/{{routeSlug}}`
- The generic project page will then act as a catchall proxy, rather than trying to do route validation
  - We let the specific project page handle the validation instead
  - If the user typed in `/project/_/memes`, each card will link to `/project/{{ref}}/memes` and that will lead to a 404 if clicked on
- Added redirects to top level pages to the intended entry page
  - e.g going to `/project/{{ref}}/auth` will redirect to `/project/{{ref}}/auth/users`
  - applied for auth, database, settings

let me know what you think @thorwebdev

image

**Comment by MildTomato** [comment]
@joshenlim looks great. ship it!

**Comment by joshenlim** [comment]
actually i'm gonna change a little on the redirect logic to shift it to next.config as @DukeManh suggested - a lot cleaner

**Review by phamhieu — APPROVED** [review]
lgtm!!! 

I add small fix to support redirect to generic links after auto authentication. 
```
// the url needs to be in this format.
// When user clicks on the link and not login yet, it will auto trigger github login
http://localhost:8082/?next=project/_/database


// if user is logout, this url will redirect to landing page
http://localhost:8082/project/_/database
```

**Pull request merged**